File: code/backend/app/models/slot.py
from datetime import date, datetime, time, timezone
from typing import Optional
from sqlmodel import Field, SQLModel
from app.models.enums import SlotStatus


class Slot(SQLModel, table=True):
    __tablename__ = "slots"

    id: Optional[int] = Field(default=None, primary_key=True)
    equipment_id: int = Field(foreign_key="equipment.id", index=True)
    date: date
    start_time: time
    end_time: time
    capacity: int
    available_count: int
    booking_cutoff_at: datetime = Field(index=True)
    allocation_run_at: Optional[datetime] = Field(default=None)
    status: SlotStatus = Field(default=SlotStatus.OPEN, index=True)
    created_by: int = Field(foreign_key="users.id")
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

File: code/web_dashboard/src/types/slot.ts
// Mirrors backend app/schemas/slot.py exactly.

export type SlotStatus = 'OPEN' | 'PENDING_ALLOCATION' | 'FULL' | 'CLOSED';

export interface Slot {
  id: number;
  equipment_id: number;
  equipment_name: string | null;
  date: string;
  start_time: string;
  end_time: string;
  capacity: number;
  available_count: number;
  booking_cutoff_at: string;
  allocation_run_at: string | null;
  status: SlotStatus;
  created_at: string;
}

export interface SlotCreateRequest {
  equipment_id: number;
  date: string;
  start_time: string;
  end_time: string;
  capacity: number;
  booking_cutoff_at: string;
}

export interface SlotUpdateRequest {
  date?: string;
  start_time?: string;
  end_time?: string;
  capacity?: number;
  booking_cutoff_at?: string;
}

export interface AllocationResult {
  slot_id: number;
  allocation_run_at: string | null;
  confirmed_count: number;
  waitlisted_count: number;
  slot_status: SlotStatus;
}

export interface DeactivateResult {
  id: number;
  status: SlotStatus;
  cancelled_bookings: number;
}
